### PR TITLE
feat(medusa): Convert IdempotencyKeyService to TypeScript

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

Migrate the service to TS

Fix the lock method which was not awaiting the retrieve method, which can lead to no lock and therefore no conflict where a conflict should occurs

FIXES CORE-349

### CODE REVIEW

File: packages/medusa/src/services/idempotency-key.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        )
+
+        const data: DeepPartial<IdempotencyKey> = {
+          recovery_point: recovery_point ?? "finished",
+        }
+
+        if (!recovery_point) {
+          data.response_body = response_body
+          data.response_code = response_code
+        }
```
Comment (pKorsholm) [inline packages/medusa/src/services/idempotency-key.ts:186]: praise: this is a neat way of handling this control flow

**Review by pKorsholm — APPROVED** [review]
lgtm

File: packages/medusa/src/services/idempotency-key.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    reqPath: string
+  ): Promise<IdempotencyKey> {
+    return await this.atomicPhase_(async () => {
+      return (
+        (await this.retrieve(headerKey).catch(() => void 0)) ??
+        (await this.create({
+          request_method: reqMethod,
+          request_params: reqParams,
+          request_path: reqPath,
+        }))
```
Comment (olivermrbl) [inline packages/medusa/src/services/idempotency-key.ts:50]: thought: I believe we are compromising readability just to use the coalescing operator here. Would rather have a method body easy to comprehend at first glance. What do you think?

File: packages/medusa/src/services/idempotency-key.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+   */
+  async create(payload: CreateIdempotencyKeyInput): Promise<IdempotencyKey> {
+    return await this.atomicPhase_(async (manager) => {
+      const idempotencyKeyRepo = manager.getCustomRepository(
+        this.idempotencyKeyRepository_
+      )
+
+      payload.idempotency_key = payload.idempotency_key ?? v4()
+
+      const created = await idempotencyKeyRepo.create(payload)
```
Comment (olivermrbl) [inline packages/medusa/src/services/idempotency-key.ts:70]: todo: Remove `await`
```suggestion
      const created = idempotencyKeyRepo.create(payload)
```

File: packages/medusa/src/services/idempotency-key.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    return await this.atomicPhase_(async (manager) => {
+      const idempotencyKeyRepo = manager.getCustomRepository(
+        this.idempotencyKeyRepository_
+      )
+
+      const key = await this.retrieve(idempotencyKey)
+
+      const isLocked =
+        key.locked_at &&
+        new Date(key.locked_at).getTime() > Date.now() - KEY_LOCKED_TIMEOUT
```
Comment (olivermrbl) [inline packages/medusa/src/services/idempotency-key.ts:114]: question: Curious, why did you choose to handle the comparison this way? Is something wrong with the initial approach?

File: packages/medusa/src/services/idempotency-key.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+   * @param idempotencyKey - key to lock
+   * @return result of the update operation
+   */
+  async lock(idempotencyKey: string): Promise<IdempotencyKey | never> {
+    return await this.atomicPhase_(async (manager) => {
+      const idempotencyKeyRepo = manager.getCustomRepository(
+        this.idempotencyKeyRepository_
+      )
+
+      const key = await this.retrieve(idempotencyKey)
```
Comment (olivermrbl) [inline packages/medusa/src/services/idempotency-key.ts:110]: praise: Nice catch! 💪

File: packages/medusa/src/services/idempotency-key.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+      const isLocked =
+        key.locked_at &&
+        new Date(key.locked_at).getTime() > Date.now() - KEY_LOCKED_TIMEOUT
+
+      if (isLocked) {
+        throw new MedusaError("conflict", "Key already locked")
```
Comment (olivermrbl) [inline packages/medusa/src/services/idempotency-key.ts:117]: suggestion: Create a new error type `CONFLICT` similar to how we have `NOT_FOUND` etc.

**Review by olivermrbl — COMMENTED** [review]
Great work! Have left a couple of comments and thoughts

File: packages/medusa/src/services/idempotency-key.ts (hunk 4 above)
Comment (adrien2p) [reply packages/medusa/src/services/idempotency-key.ts:114]: We were comparing a date agains millisecond which is a number and created a type issue. It works, but just resolve the type issue. we are now comparing milliseconds agains milliseconds (number against number)

File: packages/medusa/src/services/idempotency-key.ts (hunk 6 above)
Comment (adrien2p) [reply packages/medusa/src/services/idempotency-key.ts:117]: 👌

File: packages/medusa/src/services/idempotency-key.ts [hunk 7]
```diff
@@ -114,7 +115,7 @@ class IdempotencyKeyService extends TransactionBaseService<IdempotencyKeyService
         new Date(key.locked_at).getTime() > Date.now() - KEY_LOCKED_TIMEOUT
 
       if (isLocked) {
-        throw new MedusaError("conflict", "Key already locked")
+        throw new MedusaError(MedusaError.Types.CONFLICT, "Key already locked")
```
Comment (olivermrbl) [inline packages/medusa/src/services/idempotency-key.ts:118]: todo: Add case to `error-handler.ts` and respond with 409 similar to the transactional errors

File: packages/medusa/src/api/middlewares/error-handler.ts [hunk 8]
```diff
@@ -38,6 +38,12 @@ export default () => {
         errObj.message =
           "The request conflicted with another request. You may retry the request with the provided Idempotency-Key."
         break
+      case MedusaError.Types.CONFLICT:
+        statusCode = 409
+        errObj.code = MedusaError.Types.CONFLICT
+        errObj.message =
+          "The request conflicted with another request. You may retry the request with the provided Idempotency-Key."
+        break
```
[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa/src/api/middlewares/error-handler.ts:46]: I believe it's reasonable to respond with `code = INVALID_STATE_ERROR` for conflicts i.e. we don't need a condition here. What do you think?

[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/api/middlewares/error-handler.ts:46]: yeah you are right. It is reasonable 👍

**Review by olivermrbl — APPROVED** [review]
LGTM!

**Pull request merged**